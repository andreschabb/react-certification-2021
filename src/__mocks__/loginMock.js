export const loginModalOpened = {
  open: true,
  setOpen: jest.fn(),
};

export const loginModalClosed = {
  open: false,
  setOpen: jest.fn(),
};
