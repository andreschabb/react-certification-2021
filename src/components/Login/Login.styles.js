import styled from 'styled-components';

export const LoginForm = styled.form`
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const ModalTitle = styled.h2`
  text-transform: capitalize;
  font-weight: normal;
  text-align: center;
`;

export const FormGroup = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;
`;

export const InputLabel = styled.strong`
  display: block;
  font-weight: 700;
  text-transform: capitalize;
  margin-bottom: 0.4rem;
`;

export const LoginInput = styled.input`
  color: black;
  font-size: 1.2rem;
  width: 100%;
  padding: 0.4rem 0.6rem;
  border-radius: 3px;
  border: 1px solid black;
  background-color: white;

  &:focus {
    outline: none !important;
  }
`;

export const SubmitButton = styled.button`
  width: 7rem;
  margin-top: 1rem;
  padding: 0.4rem 0.6rem;
  font-size: 1rem;
  border-radius: 3px;
  border: 1px solid green;
  background-color: white;
  transition: 0.4s;
  color: green;
  cursor: pointer;

  &:hover {
    font-size: 1.3rem;
  }
`;
