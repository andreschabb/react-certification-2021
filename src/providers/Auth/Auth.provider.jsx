import React, { useState, useEffect, useContext, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const mockedUser = {
  id: '123',
  name: 'Wizeline',
  avatarUrl:
    'https://media.glassdoor.com/sqll/868055/wizeline-squarelogo-1473976610815.png',
};

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [user, setUser] = useState({});

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setUser(lastAuthState);
    setAuthenticated(isAuthenticated);
  }, []);

  const login = useCallback(async (username, password) => {
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        if (username === 'wizeline' && password === 'Rocks!') {
          setAuthenticated(true);
          setUser(mockedUser);
          storage.set(AUTH_STORAGE_KEY, mockedUser);
          return resolve(mockedUser);
        }
        return reject(new Error('Username or password invalid'));
      }, 500);
    });
  }, []);

  const logout = useCallback(() => {
    setUser(null);
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, null);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, user }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth };
export default AuthProvider;
