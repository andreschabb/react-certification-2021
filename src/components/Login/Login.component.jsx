import React, { useState } from 'react';
import { Modal } from 'semantic-ui-react';

import { useAuth } from '../../providers/Auth';
import {
  ErrorLog,
  FormGroup,
  InputLabel,
  LoginForm,
  LoginInput,
  ModalTitle,
  SubmitButton,
} from './Login.styles';

const Login = ({ open, setOpen }) => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');
  const { login } = useAuth();

  const handleClose = () => {
    setUsername('');
    setPassword('');
    setOpen(false);
  };

  const authenticate = async (e) => {
    e.preventDefault();
    try {
      await login(username, password);
      handleClose();
    } catch (err) {
      setError(err.message);
      setTimeout(() => {
        setError(null);
      }, 3000);
    }
  };

  return (
    <Modal
      data-testid="login-modal"
      closeIcon
      dimmer="blurring"
      size="mini"
      onClose={handleClose}
      open={open}
    >
      <Modal.Header>
        <ModalTitle>welcome back!</ModalTitle>
      </Modal.Header>
      <Modal.Content>
        <LoginForm onSubmit={authenticate}>
          <FormGroup>
            <InputLabel>username: </InputLabel>
            <LoginInput
              required
              type="text"
              data-testid="login-username-input"
              id="username"
              value={username}
              onChange={(event) => setUsername(event.target.value)}
            />
          </FormGroup>
          <FormGroup>
            <InputLabel>password: </InputLabel>
            <LoginInput
              required
              type="password"
              data-testid="login-password-input"
              id="password"
              value={password}
              onChange={(event) => setPassword(event.target.value)}
            />
          </FormGroup>
          {error && <ErrorLog>{error}</ErrorLog>}
          <SubmitButton data-testid="login-button" type="submit">
            Login
          </SubmitButton>
        </LoginForm>
      </Modal.Content>
    </Modal>
  );
};

export default Login;
