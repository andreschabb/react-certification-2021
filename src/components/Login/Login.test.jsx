import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { loginModalClosed, loginModalOpened } from '../../__mocks__/loginMock';
import { useAuth } from '../../providers/Auth';
import Login from './Login.component';

const mockLogin = jest.fn();

jest.mock('../../providers/Auth');

describe('Login', () => {
  beforeEach(() => {
    useAuth.mockImplementation(() => ({ authenticated: false }));
  });

  it('should renders when modal is opened', () => {
    const { container, getByTestId } = render(<Login {...loginModalOpened} />);

    expect(getByTestId('login-modal')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should not renders when modal is closed', () => {
    const { queryByTestId } = render(<Login {...loginModalClosed} />);

    expect(queryByTestId('login-modal')).not.toBeInTheDocument();
  });

  it('should allows user to login', async () => {
    useAuth.mockImplementation(() => ({ authenticated: true, login: mockLogin }));
    const { getByTestId } = render(<Login {...loginModalOpened} />);

    const usernameInput = getByTestId('login-username-input');
    const passwordInput = getByTestId('login-password-input');

    fireEvent.change(usernameInput, { target: { value: 'admin' } });
    fireEvent.change(passwordInput, { target: { value: 'admin' } });
    expect(usernameInput.value).toBe('admin');
    expect(passwordInput.value).toBe('admin');

    await fireEvent.click(getByTestId('login-button'));
    expect(usernameInput.value).toBe('');
    expect(passwordInput.value).toBe('');

    expect(mockLogin).toHaveBeenCalledTimes(1);
  });
});
