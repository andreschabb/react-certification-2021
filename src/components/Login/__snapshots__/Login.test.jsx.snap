// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should renders when modal is opened 1`] = `<div />`;
